README.md
# cmsc414buildit
REFER TO BUILD DIRECTORY FOR CODE

Potential attacks: 
Man in the middle (Impersonating client/server maybe?)
Repeat attacks (Listening to and repeating old messages to make an effect happen)
Eavesdropping (Listening and understanding communications)
Buffer overflows (giving atm lots of input that is bad input)
Denial of Service (not failing server when it gets a bad input)
Password guessing attack 
Integer overflow 
File name attacks 
:D
OwO

FIXED Known bug: on "card file already exists" or "card file does not exist", bank crashes. 
FIXED Known bug: on "account for deposit does not exist", bank crashes

changelog.txt
6:30pm 12/2 Clark- Added basic functionality to understand and validate command line arguments in bank-main.c.
9:44pm 12/2 Clark- Added basic functionality to parse but not validate command line arguments in atm-main.c.
10:22pm 12/2 Clark- Increased the functionality to parse command line arguments in atm-main.c. 
8:11pm 12/3 Clark- Added functionality to send information about the commands from atm to bank.
12:34am 12/4 Clark- Symetrically encrypting from atm->bank but not decrypting. 
12:12am 12/5 Clark- Decrypting at bank but having some small issues parsing the result. 
2:18am (fml) 12/8 Matthew - Converting string to ull for account balance and adding some int-overflow countermeasures. 
    Create, deposit, and withdraw updated (withdraw still untested i gotta go to bed man D:). 
    Basic functionality works but I can make it better (decimals are messed up rn)
11:45pm 12/10 Clark- Did the arg parsing fixes and started the repeat attack protection
4:23pm 12/11 Clark- Finished repeat attack protection.
2:34am 12/12 Matthew -- Finished behavior for different commands in bank-main.c. Can confirm it works by output of
    printLinkedList(), but still need to work on the final_value string being accurate to the last transaction made. this 
    shouldn't be too tough tho. I'll go back and wrap that up either tomorrow or monday
2:15pm 12/13 Matthew -- Finished construction of output string for new accounts created -- TODO still need to Converting
    sent_value_of_operation to cents instead of dollars and update everything for that accordingly
2:38am 12/14 Matthew -- FINISHED construction of the output string depending on transaction type, as well as remaining basic 
    functionality for each command (hype). STILL TODO VERY IMPORTANT -- I moved the int overflow protection from bank-main to
    atm-main, to avoid the network socket not closing issue, but now have to find out about how Clark parsed everything in that
    program ("parsing hell"). Not looking forward to that but still hype that it WORKS :D I gotta stop staying up this late
2:02am 12/15 Matthew -- Got a working solution to simple int overflow attack via negative input -- will look further into
    whether or not it is the permanent (or only) one. I think it is pretty much done, if not nearly so. neat :D
TOO LATEam 12/16 Matthew -- I think it is pretty much done now -- I made sure it was behaving correctly in the event of invalid
    or bad commands as well as when bank-main does not pass the repeat attack test. The only thing that I can think of left is
    to find a way to implement the timer for ATM to close the connection if it does not hear from bank in long enough, since
    people on Campuswire are saying that that breaks the auto-tests.

bank.c
#include "bank.h"
#include "net.h"
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <openssl/rand.h>

Bank* bank_create(char *auth_file, char *ip, unsigned short port)
{

	Bank *bank = (Bank*) calloc(1, sizeof(Bank));
        
	if(bank == NULL) {
		perror("Could not allocate Bank");
		exit(255);
	}

#define BOOL_CHK(x,msg) if (x) { perror(msg); exit(255); }


	/* setup network connection */
	BOOL_CHK(inet_pton(AF_INET, ip, &(bank->bank_addr.sin_addr)) != 1, "could not convert ip address");

	bank->bank_addr.sin_port = htons(port);
	bank->bank_addr.sin_family = AF_INET;

	int s = socket(AF_INET, SOCK_STREAM, 0);
	BOOL_CHK(s<0,"could not create socket");

	int enable = 1;
	BOOL_CHK(setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int)) < 0, "setsockopt(SO_REUSEADDR) failed");

	BOOL_CHK(bind(s, (struct sockaddr*)&(bank->bank_addr), sizeof(bank->bank_addr)) < 0, "could not bind");

	listen(s, 5);

	bank->sockfd = s;
	struct timeval timeout;      
    timeout.tv_sec = 100;
    timeout.tv_usec = 0;

    if (setsockopt (bank->sockfd, SOL_SOCKET, SO_RCVTIMEO, (char *)&timeout,
                sizeof(timeout)) < 0)
        exit(255);

    if (setsockopt (bank->sockfd, SOL_SOCKET, SO_SNDTIMEO, (char *)&timeout,
                sizeof(timeout)) < 0)
        exit(255);

#undef BOOL_CHK



	// Set up the protocol state
	// TODO set up more, as needed
	fflush(stdout);
	printf("created\n");
	fflush(stdout);
	return bank;
}

void bank_free(Bank *bank){

	close(bank->sockfd);
	free(bank);
	// TODO
}

/* sends data_len bytes from data to atm, returns size 0 on success, negative on failure */
int bank_send(Bank *bank, const char *data, size_t data_len) {
	if (bank->clientfd < 0) {
		return -1;
	}

	if (send_all(bank->clientfd, (const char*)&data_len, sizeof(data_len)) != sizeof(data_len)) {
		return -2;
	}

	if (send_all(bank->clientfd, data, data_len) != (ssize_t)data_len) {
		return -3;
	}
	return 0;	
}

/* receive a message (i.e., something sent via atm_send) and store it
 * in data. If the message exceeds max_data_len, a negative value is
 * returned and the message is discarded */
ssize_t bank_recv(Bank *bank, char *data, size_t max_data_len) {

	size_t msg_len;

	if (bank->clientfd < 0) {
		return -1;
	}

	if (recv_all(bank->clientfd, (char*)&msg_len, sizeof(msg_len)) != sizeof(msg_len)) {
		return -2;
	}

	if (msg_len > max_data_len) {
		/* message doesn't fit in data, read all of the message to discard it */
		char tmp[4096];
		do {
			size_t to_read = msg_len > sizeof(tmp) ? sizeof(tmp) : msg_len;
			if (recv_all(bank->clientfd, tmp, to_read) != sizeof(to_read)) {
				/* logic error somewhere, should probably crash/restart */
				return -3;
			}
			msg_len -= to_read;
		} while(msg_len > 0) ;
	}

	return recv_all(bank->clientfd, data, msg_len);	

}

void bank_process_remote_command(Bank *bank, char *command, size_t len) {
	// TODO: Implement the bank side of the ATM-bank protocol
}

//Add more functions as required


bank-main.c
#include "bank.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/ssl.h>
#include <openssl/evp.h>
#include <openssl/bio.h>
#include <openssl/conf.h>
#include <openssl/err.h>

// Default port and ip address are defined here

int sym_encrypt(unsigned char *plaintext, int plaintext_len, unsigned char *key, unsigned char *iv, unsigned char *ciphertext){
	EVP_CIPHER_CTX *ctx;
	int len;
	int ciphertext_len;

	// creat and init context
	if(!(ctx = EVP_CIPHER_CTX_new())){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 16\n");
		exit(255);
	}

	// init encryption operation 
	if(1!= EVP_EncryptInit_ex(ctx,EVP_aes_256_cbc(),NULL,key,iv)){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 22\n");
		exit(255);
	}

	// provide message to be encrypted/ get output
	if(1!= EVP_EncryptUpdate(ctx,ciphertext,&len,plaintext,plaintext_len)){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 28\n");
		exit(255);
	}
	ciphertext_len = len;

	// finalize the encryption 
	if(1!=EVP_EncryptFinal_ex(ctx,ciphertext+len,&len)){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 28\n");
		exit(255);
	}
	ciphertext_len += len;

	// clean up and return 
	EVP_CIPHER_CTX_free(ctx);
	return ciphertext_len;
}

int sym_decrypt(unsigned char *ciphertext, int ciphertext_len, unsigned char *key, unsigned char *iv, unsigned char *plaintext){
	EVP_CIPHER_CTX *ctx;
	int len;
	int plaintext_len;

	// create and init context
	if(!(ctx = EVP_CIPHER_CTX_new())){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 20\n");
		exit(255);
	}

	// init decrypt operation
	if(1!= EVP_DecryptInit_ex(ctx,EVP_aes_256_cbc(),NULL,key,iv)){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 26\n");
		exit(255);
	}

	// put decrypted message, get output
	if(1!= EVP_DecryptUpdate(ctx,plaintext,&len,ciphertext,ciphertext_len)){
		// printf("DEBUG: Wuhwoh things are getting fonky at line 32\n");
		exit(255);
	}
	plaintext_len = len;

	//Cleanup 
	EVP_CIPHER_CTX_free(ctx);
	return plaintext_len;
}
#define ull unsigned long long
struct linked_list_node {
	ull account_balance;
	char *account; // key
	// char *balance_str;
	unsigned char *card_info;
	struct linked_list_node *next;
};

struct linked_list_node *head = NULL;
struct linked_list_node *curr = NULL;

// Converts ull to string
char * utoa(unsigned int n)
{
  char * res, buf[30]; // long enough for largest number
  unsigned int i, counter = 0;

  if (n == 0)
    buf[counter++] = '0';

  for ( ; n; n /= 10)
    buf[counter++] = "0123456789"[n%10];

  res = malloc(counter);

  for (i = 0; i < counter; ++i)
    res[i] = buf[counter - i - 1];
  res[i] = '\0';
  return res;
}

void insert(char *account,char *account_balance,unsigned char *card_info){
	int neg_check = atoi(account_balance);
	// printf("DEBUG: Received initial bal %s\n", account_balance);
	if (neg_check < 0) {
		// printf("ERROR: Please deposit a positive number\n");
		exit(255);
	}
	ull bal = (ull) atoi(account_balance);
	if (bal >= 4294967295.99) {
		// printf("ERROR: Value to deposit too big -- please deposit in smaller increments\n");
		exit(255);
	}
	struct linked_list_node *new_node = (struct linked_list_node*)malloc(sizeof(struct linked_list_node));
	
	new_node->account = malloc(strlen(account)+1);
	// memset(new_node->account,'\0',123);
	strcpy(new_node->account,account);
	// strcpy(new_node->balance_str, account_balance);

	new_node->account_balance = (ull) 0;
	// memset(new_node->account_balance,'\0',15);
	new_node->account_balance += bal;

	new_node->card_info = malloc(strlen(card_info)+1);
	// memset(new_node->card_info,'\0',33);
	strcpy(new_node->card_info,card_info);

	new_node->next = head;
	head = new_node;
}

void printLinkedList(){
	struct linked_list_node *curr = head;
	printf("\n[ ");

	while(curr != NULL){
		printf("(account:%s,\n  account_balance:%u,\n  card_info:%s)\n",curr->account,curr->account_balance ,curr->card_info);
		curr = curr->next;
	}

	printf(" ]\n");
}

struct linked_list_node* find_account(char  *account){
	struct linked_list_node* curr = head;
	if(head==NULL){
		return NULL;
	}
	while(strcmp(curr->account,account)!=0){
		if(curr->next == NULL){
			return NULL;
		}else{
			curr = curr->next;
		}
	}
	return curr;
}

int main(int argc, char** argv){
	int i = 1;
	char *ip = "127.0.0.1";
	FILE *file;
	int auth_pos = -1; // returns the location in argv of an auth file. If not specified, it is -1.

	int recval_locs[] = {-1,-1};
	while(i<argc){
		if(strcmp(argv[i],"-p")==0){
			// printf("DEBUG: %s argument detected at %d with %s value\n",argv[i],i,argv[i+1]);
			if(recval_locs[0]!=-1){
				// printf("DEBUG: Duplicate %s argument detected at %d with %s value\n",argv[i],i,argv[i+1]);
				exit(255);
			}else if(i+1>=argc){
				// printf("DEBUG: Argument %s detected at %d has no value\n",argv[i],i);
				exit(255);
			}else{
				recval_locs[0] = i;
			}
			i+=2;
		}else if(strcmp(argv[i],"-s")==0){
			// printf("DEBUG: %s argument detected at %d with %s value\n",argv[i],i,argv[i+1]);
			if(recval_locs[1]!=-1){
				// printf("DEBUG: Duplicate %s argument detected at %d with %s value\n",argv[i],i,argv[i+1]);
				exit(255);
			}else if(i+1>=argc){
				// printf("DEBUG: Argument %s detected at %d has no value\n",argv[i],i);
				exit(255);
			}else{
				recval_locs[1] = i;
			}
			i+=2;
		}else if(strlen(argv[i])>2&&argv[i][0]=='-'&&argv[i][1]=='p'){
			// printf("DEBUG: %c argument at %d was a nospace attempt.\n",argv[i][1],i);
			if(recval_locs[0]!=-1){
				// printf("DEBUG: Duplicate %c argument detected at %d with %s value.\n",argv[i][1],i,"TODO");
				exit(255);
			}else{
				recval_locs[0] = -1-i;
			}
			i+=1;
		}else if(strlen(argv[i])>2&&argv[i][0]=='-'&&argv[i][1]=='s'){
			// printf("DEBUG: %c argument at %d was a nospace attempt.\n",argv[i][1],i);
			if(recval_locs[1]!=-1){
				// printf("DEBUG: Duplicate %c argument detected at %d with %s value.\n",argv[i][1],i,"TODO");
				exit(255);
			}else{
				recval_locs[1] = -1-i;
			}
			i+=1;
		}else{
			// printf("DEBUG: Argument at %d was invalid. Value is %s\n",i,argv[i]);
			exit(255);
		}
	}
	
	int port_num;
	unsigned short port;
	if(recval_locs[0]<-1){
		// printf("Preparing to look for port at %d\n",(recval_locs[0]+1)*-1);
		port_num = atoi(&(argv[(recval_locs[0]+1)*-1][2]));
	}else{
		port_num = (recval_locs[0]==-1 ? 3000 : atoi(argv[recval_locs[0]+1]));
	} 
	port = (unsigned short)port_num;

	char *auth_file_name;
	if(recval_locs[1]<-1){
		// printf("Preparing to look for authfile at %d\n",(recval_locs[1]+1)*-1);
		auth_file_name = &(argv[(recval_locs[1]+1)*-1][2]);
	}else{
		auth_file_name = (recval_locs[1]==-1 ? "bank.auth" : argv[recval_locs[1]+1]);
	}
	
	unsigned char sym_key[32];

	// print the values that we learned 
	// printf("DEBUG: The port, as entered, is %d\n",(int)port);
	// printf("DEBUG: The auth file is %s\n",auth_file_name);

	/* no error checking is done on any of this. may need to modify this */
	

	// if not auth_file exists, create auth_file
	//bank->auth_file = auth_file
	
	FILE *auth_file;
	if((auth_file = fopen(auth_file_name,"r"))){
		fclose(auth_file);
		// printf("DEBUG: auth file already exists. Exiting with error.\n");
		exit(255);
	}else{
		if((auth_file = fopen(auth_file_name,"w"))){
			// printf("DEBUG: writing super secret code to new auth file.\n");
			RAND_bytes(sym_key,32);
			fwrite(sym_key,1,sizeof(sym_key),auth_file);
			fclose(auth_file);
		}
	}

	Bank *b = bank_create(auth_file_name, ip, port);

	/* process each incoming client, one at a time, until complete */
	for(;;) {

		unsigned int len = sizeof(b->remote_addr);
		b->clientfd = accept(b->sockfd, (struct sockaddr*)&b->remote_addr, &len);
		if (b->clientfd < 0) {
			perror("error on accept call");
			exit(255);
		}

		

		/* okay, connected to bank/atm. Send/recv messages to/from the bank/atm. */
		char buffer[1024];
		
		bank_recv(b, buffer, sizeof(buffer));

		
		// decrypt here

		// first 16 bytes are the unsigned char iv
		unsigned char iv[16];
		unsigned char encrypted_msg[300];
		unsigned char decrypted_msg[300];
		// printf("DEBUG: Received message containing:\n(");
		for(int i=0;i<16;i++){
			//printf("%c",buffer[i]);
			iv[i] = buffer[i];
		}
		// printf(" iv\n");
		for(int i=16;i<192;i++){
			//printf("%c",buffer[i]);
			encrypted_msg[i-16] = buffer[i];
		}
		// printf(" msg)\n");
		int decrypted_length = sym_decrypt(encrypted_msg,192-16,sym_key,iv,decrypted_msg);
		// printf("DEBUG: length of decrypted message is %d.\nMessage is:(\n",decrypted_length);
		for(int i=0;i<decrypted_length+9;i++){
			//printf("%c",decrypted_msg[i]);
		}
		// printf(")\n");
		
		/* Buffer message format:
		0 account name (122 characters)
		1 card file value (32 characters)
		2 mode of operation (1 character)
		3 value of operation (13 characters)
		*/
		int buffer_idx = 0;
		char sent_account[123];
		// memset(sent_account,'\0',sizeof(sent_account));
		// strncpy(sent_account,buffer,122);
		for(int i=0;i<122;i++){
			sent_account[i] = decrypted_msg[buffer_idx++];
		}
		sent_account[122] = '\0';
		// printf("DEBUG: ATM sent an account with account name of %s.\n",sent_account);
		// printf("DEBUG: Remaining string is %s.\n",&(buffer[122]));

		unsigned char sent_card_value[33];
		// memset(sent_card_value,'\0',sizeof(sent_card_value));
		// strncpy(sent_card_value,&(buffer[122]),32);
		// printf("DEBUG: ATM sent a card value of: '");
		for(int i=0;i<32;i++){
			sent_card_value[i] = decrypted_msg[buffer_idx++];
			// printf("%c",sent_card_value[i]);
		}
		sent_card_value[32] = '\0';
		// printf("'.\n");
		// printf("DEBUG: ATM sent a card value of %s.\n",sent_card_value);
		// printf("DEBUG: Remaining string is %s.\n",&(buffer[122+32]));

		char sent_mode_of_operation[2];
		// memset(sent_mode_of_operation,'\0',sizeof(sent_mode_of_operation));
		// strncpy(sent_mode_of_operation,&(buffer[122+32]),2);
		for(int i=0;i<1;i++){
			sent_mode_of_operation[i] = decrypted_msg[buffer_idx++];
		}
		sent_mode_of_operation[2] = '\0';
		// printf("DEBUG: ATM sent a mode of operation of %s.\n",sent_mode_of_operation);
		// printf("DEBUG: Remaining string is %s.\n",&(buffer[122+32+2]));

		char sent_value_of_operation[15];
		// memset(sent_value_of_operation,'\0',sizeof(sent_value_of_operation));
		// strncpy(sent_value_of_operation,&(buffer[122+32+2]),13);
		for(int i=0;i<13;i++){
			sent_value_of_operation[i] = decrypted_msg[buffer_idx++];
		}
		sent_value_of_operation[13] = '\0';
		char *p = strchr(sent_value_of_operation, '.');
		// THIS GETS THE AMOUNT OF MONEY AS AN EXPRESSION OF CENTS -- WILL FINISH LATER
		*(p) = *(p + 1);
		*(p + 1) = *(p + 2);
		*(p + 2) = '\0';
		// printf("DEBUG: ATM sent a value of operation of %s.\n",sent_value_of_operation);
		// printf("DEBUG: Remaining string (corresponding to the anti_repeat value) is %s.\n",&(buffer[122+32+1+13]));

		
		// do some validation ?
		// TODO validation

		// if data validates: 

		// TODO: Check if the card for the user validates 
		// TODO: Implement a system to check 
		// printf("DEBUG: Communication is valid so far- defending against repeat attacks.\n");
		// do everything
		/* Buffer message format:
			0 account name (122 characters)
			1 card file value (32 characters)
			2 mode of operation (1 character)
			3 value of operation (13 characters)
		*/
		char buffer_resp[300] = "";
		// created from msg above: char sent_account[123];
		// to be filled in with random bytes: unsigned char card_rand_bytes[32];
		unsigned char card_rand_bytes[32];
		RAND_bytes(card_rand_bytes,32);

		// value doesnt matter
		char operation_value[14];
		memset(operation_value,'\0',sizeof(operation_value));
		for(int i=strlen(operation_value);i<13;i++){
			strcat(operation_value," ");
		}

		int buffer_idx_resp = 0;
		// always 122 characters
		for(int i=0;i<122;i++){
			buffer_resp[buffer_idx_resp++] = sent_account[i];
		}
		// always 32 wacky characters
		for(int i=0;i<32;i++){
			buffer_resp[buffer_idx_resp++] = card_rand_bytes[i];
		}
		// placeholder value
		// for(int i=0;i<strlen("X");i++){
		// 	buffer_resp[buffer_idx_resp++] = 'X';
		// }
		// 1 char

		// printf("Operation value is equal to '%s'.\nIt is %d characters long.\n",operation_value,strlen(operation_value));
		// strcat(buffer_resp,operation_value);
		// for(int i=0;i<strlen(operation_value);i++){
		// 	buffer_resp[buffer_idx_resp++] = operation_value[i];
		// }
		// 13 chars

		// printf("DEBUG: Printing entire message:(\n");
		// for(int i=0;i<buffer_idx_resp;i++){
		// 	printf("%c",buffer_resp[i]);
		// }
		// printf(")\n");

		// encrypt here 
		unsigned char *ciphertext = malloc(300*sizeof(char*));
		// unsigned char iv[16]; is read in above
		// unsigned char sym_key[32]; is read in above

		int ciphertext_len = sym_encrypt(buffer_resp, strlen((char*)buffer_resp),sym_key,iv,ciphertext);

		unsigned char *msg_resp = malloc(300*sizeof(char*)+16);
		for(int i=0;i<16;i++){
			msg_resp[i] = iv[i];
		}
		for(int i=16;i<16+ciphertext_len;i++){
			msg_resp[i] = ciphertext[i-16];
		}
		int msg_resp_len = ciphertext_len+16;

		// printf("DEBUG: Preparing to send message size %d containing:\n(",msg_resp_len);
		// for(int i=0;i<16;i++){
		// 	printf("%c",msg_resp[i]);
		// }
		// printf(" iv\n");
		// for(int i=16;i<msg_resp_len;i++){
		// 	printf("%c",msg_resp[i]);
		// }
		// printf(" msg)\n");
		bank_send(b, msg_resp, msg_resp_len);

		char buffer_rec2[1024];
		bank_recv(b, buffer_rec2, sizeof(buffer_rec2));
		// for(int i=0;i<208;i++){
		// 	printf("%c",buffer_rec2[i]);
		// }
		// printf(" msg");
		unsigned char decrypted_msg_rec2[300];
		int decrypted_length_rec2 = sym_decrypt(buffer_rec2,208,sym_key,iv,decrypted_msg_rec2);
		// printf("DEBUG: Received decrypted message of size %d containing:(\n",decrypted_length_rec2);
		// for(int i=0;i<122;i++){
		// 	printf("%c",decrypted_msg_rec2[i]);
		// }
		// printf(" acct\n");
		int repeat = 0;
		for(int i=122;i<122+16;i++){
			//printf("%c",decrypted_msg_rec2[i]);
			if(card_rand_bytes[i-122]!=decrypted_msg_rec2[i]){
				// printf("(%c!=%c)",card_rand_bytes[i-122],decrypted_msg_rec2[i]);
				// TODO close connection
				// printf("DEBUG: Repeat attack detected! Terminating the connection.\n");
				repeat = 1;
			}
		}
		// printf(" val\n");
		// printf("DEBUG: Expected :(\n");
		// for(int i=0;i<16;i++){
		// 	printf("%c",card_rand_bytes[i]);
		// }
		// printf(" value\n");

		// Verified that this is not a repeat attack
		// printf("DEBUG: Everything so far is valid- changes to state are now going to be made and returned.\n");
				int is_valid = 1;
		struct linked_list_node *acc;
		char final_value[300] = "{\"account\":\"";
		strcat(final_value, sent_account);
		if (repeat) {
			strcat(final_value, "INVALID");
		} else {
			if('n'==sent_mode_of_operation[0]){
				// printf("DEBUG: The atm wants to make a new account.\n");
				if(find_account(sent_account)!=NULL){
					// printf("DEBUG: That account for '%s' already exists!\n",sent_account);
					is_valid = 0;
					strcat(final_value, "INVALID");
				}else{
					insert(sent_account,sent_value_of_operation,sent_card_value);
					ull val = (ull) atoi(sent_value_of_operation);
					// printf("DEBUG -- VAL = %s\n", utoa(val / 100));
					strcat(final_value, "\", \"initial_balance\":");
					strcat(final_value, utoa(val / 100));
					strcat(final_value, ".");
					if (val % 100 != 0) {
						
						strcat(final_value, utoa(val % (ull) 100));
					} else {
					 	strcat(final_value, "00");
					}
					//printLinkedList();
				}
			}else if('d'==sent_mode_of_operation[0]){
				// printf("DEBUG: The atm wants to deposit.\n");
				struct linked_list_node *found = find_account(sent_account);
				if(found==NULL){
					// printf("DEBUG: The account for '%s' does not exist!\n",sent_account);
					is_valid = 0;
					strcat(final_value, "INVALID");
				}else{
					// check if card is valid
					int valid = 1;
					for(int i=0;i<32;i++){
						if(found->card_info[i]!=sent_card_value[i]){
							valid = 0;
						}
					}
					if(valid == 1){
						acc = found;
						// printf("DEBUG: Changing found's values..... matthew started this hopefully its finished and/or working.....\n");
						// TODO move this operation to later until the atm has passed the repeat attack check
						ull val = (ull) atoi(sent_value_of_operation);
						if ((val + found->account_balance) < LLONG_MAX) {
							found->account_balance += val;
							strcat(final_value, "\", \"deposit\":");
							strcat(final_value, utoa(val / (ull) 100));
							strcat(final_value, ".");
							if (val % 100 >= 10) {
								
								strcat(final_value, utoa(val % (ull) 100));
							} else {
							 	strcat(final_value, "0");
								strcat(final_value, utoa(val % (ull) 100));
							}
							//printLinkedList();
						} else {
							// printf("ERROR -- Max account limit exceeded -- please withdraw some funds\n");
							is_valid = 0;
							strcat(final_value, "INVALID");
						}
					}else{
						// printf("DEBUG: Woah woah buddy, you dont have the right card to access that account.\n");
						is_valid = 0;
						strcat(final_value, "INVALID");
					}
				}
			}else if('w'==sent_mode_of_operation[0]){
				// printf("DEBUG: The atm wants to widthdraw.\n");
				struct linked_list_node *found = find_account(sent_account);
				if(found==NULL){
					// printf("DEBUG: The account for '%s' does not exist!\n",sent_account);
					is_valid = 0;
					strcat(final_value, "INVALID");
				}else{
					// check if card is valid
					int valid = 1;
					for(int i=0;i<32;i++){
						if(found->card_info[i]!=sent_card_value[i]){
							valid = 0;
						}
					}
					if(valid == 1){
						acc = found;
						// printf("DEBUG: Changing found's values..... implement this as soon as I know whether we have to store superlarge numbers or not.....\n");
						ull val = (ull) atoi(sent_value_of_operation);
						if ((int) (found->account_balance - val) >= 0) {
							found->account_balance -= val;
							strcat(final_value, "\", \"withdraw\":");
							strcat(final_value, utoa(val / (ull) 100));
							strcat(final_value, ".");
							if (val % 100 >= 10) {
								strcat(final_value, utoa(val % (ull) 100));
							} else {
								strcat(final_value, "0");
								strcat(final_value, utoa(val % (ull) 100));
							}
						} else {
							// printf("DEBUG -- Tried to withdraw more than you had D:\n");
							is_valid = 0;
							strcat(final_value, "INVALID");
							// TODO -- FIND A SOLUTION TO THIS USAGE ERROR
						}
						//printLinkedList();
					}else{
						// printf("DEBUG: Woah woah buddy, you dont have the right card to access that account.\n");
						is_valid = 0;
						strcat(final_value, "INVALID");
					}
				}
			}else if('g'==sent_mode_of_operation[0]){
				// printf("DEBUG: The atm wants to check the balance of an account.\n");
				struct linked_list_node *found = find_account(sent_account);
				if(found==NULL){
					// printf("DEBUG: The account for '%s' does not exist!\n",sent_account);
					is_valid = 0;
					strcat(final_value, "INVALID");
				}else{
					// check if card is valid
					int valid = 1;
					for(int i=0;i<32;i++){
						if(found->card_info[i]!=sent_card_value[i]){
							valid = 0;
						}
					}
					if(valid == 1){
						acc = found;
						// printf("DEBUG: Changing found's values..... implement this as soon as I know whether we have to store superlarge numbers or not.....\n");
						// printf("Account for %s has %u funds\n", found->account, found->account_balance);
						ull val = found->account_balance;
						strcat(final_value, "\", \"balance\":");
						strcat(final_value, utoa(val / (ull) 100));
						strcat(final_value, ".");
						if (val % 100 >= 10) {
							
							strcat(final_value, utoa(val % (ull) 100));
						} else {
							strcat(final_value, "0");
							strcat(final_value, utoa(val % (ull) 100));
						}
						//printLinkedList();
					}else{
						// printf("DEBUG: Woah woah buddy, you dont have the right card to access that account.\n");
						is_valid = 0;
						strcat(final_value, "INVALID");
					}
				}
			}else{
				// printf("DEBUG: The atm sent an invalid mode of operation: '%c'.\n",sent_mode_of_operation[0]);
				is_valid = 0;
				strcat(final_value, "INVALID");
			}
		}

		// msg asking for modified echo here
		if(is_valid==1 && repeat==0){
			strcat(final_value, "}");
			fflush(stdout);
			for (int i=0; final_value[i] != '\0'; i++) {
				if (final_value[i] != ' ') {
					printf("%c", final_value[i]);
				}
			}
			printf("\n");
			fflush(stdout);
		}

		// MATTHEW TODO MAKE CHANGES TO BANK STATE HERE- EVERYHTING IS VERIFIED NOW
		// Am going to finish this tomorrow -- want to parse this string to be accurate to last transaction
		// Shouldn't be too tough
		//char final_value[300] = "\"account\":\"bob\",\"initial_balance\":\"10.00\"\0";
		// char final_value[300] = acc->account;
		// char bal[20];
		// strcat(final_value, "\"balance\":");
		// sprintf(bal, "%u", acc->account_balance);
		// strcat(final_value, bal);
		// strcat(final_value, "\"\0");

		//printLinkedList();
		// CHANGES ARE DONE BEING MADE HERE
		// NOW PRINTING FINAL VALUE FOR BANK
		// for(int i=strlen(final_value);i<300;i++){
		// 	strcat(final_value," ");
		// }
		// for(int i=0;i<300;i++){
		// 	if(final_value[i]!=' '){
		// 		printf("%c",final_value[i]);
		// 	}
		// }
		// printf("\n");

		// always 300 characters
		unsigned char final_buffer_resp[300] = "";
		for(int i=0;i<300;i++){
			final_buffer_resp[i] = final_value[i];
		}

		unsigned char final_ciphertext[500];
		int final_ciphertext_len = sym_encrypt(final_buffer_resp, 300 ,sym_key,iv,final_ciphertext);

		unsigned char final_message[600];
		for(int i=0;i<16;i++){
			final_message[i] = iv[i];
		}
		for(int i=16;i<16+final_ciphertext_len;i++){
			final_message[i] = final_ciphertext[i-16];
		}
		int final_message_len = final_ciphertext_len+16;
		// decrypted message is 300 characters long 
		// printf("DEBUG: Final encrypted message is %d characters long.\n",final_message_len);
		bank_send(b, final_message, final_message_len);
	

		/* when finished processing commands ...*/
		close(b->clientfd);
		b->clientfd = -1;
	}


	
	
	// Implement how atm protocol will work: sanitizing inputs and using different modes of operations

	return 0;
}
